libcloud/compute/drivers/docker.py

```diff
@@ -0,0 +1,573 @@
+    import json
+
+from libcloud.utils.py3 import httplib
+from libcloud.utils.py3 import b
+
+from libcloud.compute.providers import Provider
+from libcloud.common.base import JsonResponse, ConnectionUserAndKey
+from libcloud.compute.types import (NodeState, InvalidCredsError)
+from libcloud.compute.base import (Node, NodeDriver, NodeImage,
+                                   NodeSize)
+
+VALID_RESPONSE_CODES = [httplib.OK, httplib.ACCEPTED, httplib.CREATED,
+                        httplib.NO_CONTENT]
+
+
+class DockerResponse(JsonResponse):
+
+    valid_response_codes = [httplib.OK, httplib.ACCEPTED, httplib.CREATED,
+                            httplib.NO_CONTENT]
+
+    def parse_body(self):
+        if len(self.body) == 0 and not self.parse_zero_length_body:
+            return self.body
+
+        try:
+            # error responses are tricky in Docker. Eg response could be
+            # an error, but response status could still be 200
+            body = json.loads(self.body)
+        except ValueError:
+            m = re.search('Error: (.+?)"', self.body)
+            if m:
+                error_msg = m.group(1)
+                raise Exception(error_msg)
+            else:
+                raise Exception(
+                    'ConnectionError: Failed to parse JSON response')
+        return body
+
+    def parse_error(self):
+        if self.status == 401:
+            raise InvalidCredsError('Invalid credentials')
+        return self.body
+
+    def success(self):
+        return self.status in self.valid_response_codes
+
+
+class DockerConnection(ConnectionUserAndKey):
+
+    responseCls = DockerResponse
+    timeout = 60
+
+    def add_default_headers(self, headers):
+        """
+        Add parameters that are necessary for every request
+        If user and password are specified, include a base http auth
+        header
+        """
+        headers['Content-Type'] = 'application/json'
+        if self.user_id and self.key:
+            user_b64 = base64.b64encode(b('%s:%s' % (self.user_id, self.key)))
+            headers['Authorization'] = 'Basic %s' % (user_b64.decode('utf-8'))
+        return headers
+
+
+class DockerNodeDriver(NodeDriver):
+    """
+    Docker node driver class.
+
+    >>> from libcloud.compute.providers import get_driver
+    >>> driver = get_driver('docker')
+    >>> conn = driver(host='198.61.239.128', port=4243)
+    >>> conn.list_nodes()
+    or connecting to http basic auth protected https host:
+    >>> conn = driver('user', 'pass', host='https://198.61.239.128', port=443)
+
+    connect with tls authentication, by providing a hostname, port, a private
+    key file (.pem) and certificate (.pem) file
+    >>> conn = driver(host='https://198.61.239.128',
+        port=4243, key_file='key.pem', cert_file='cert.pem')
+    """
+
+    type = Provider.DOCKER
+    name = 'Docker'
+    website = 'http://docker.io'
+    connectionCls = DockerConnection
+    features = {'create_node': ['password']}
```

Comment:
I think that the need to support docker or other ecosystem in container universe, likewise any other cloud provider.
You want to "provision" without the specifics or the underlying provider. 
As you do for: AWS, Openstack, Digital Ocean, etc...

On the Docker support, it's the same, you just need to provision. Does not matter if is Docker, Swarm, Kubernetes, or any other Container implementation.
At least, it's the way I think... and for that reason, that I looked into this with 100% interest. :smile: 

I want to create a client for Instance Provision and not worry about the provider and for my use-case docker (containers) belongs to the providers available...